cmd/validate.go

```diff
@@ -0,0 +1,275 @@
+'read', 'stream', and 'execute' methods respectively, with '-' representing lack of support for a
+method.`,
+		Args:   cobra.ExactArgs(1),
+		PreRun: bindServerArgs,
+		RunE:   toRunE(validateMain),
+	}
+	validateCmd.Flags().IntP("parallel", "p", 10, "Number of entries to validate in parallel")
+	validateCmd.Flags().BoolP("all", "a", false, "Validate all entries rather than an example at each level of hierarchy")
+	addServerArgs(validateCmd, "warn")
+	return validateCmd
+}
+
+func validateMain(cmd *cobra.Command, args []string) exitCode {
+	// Validate that 'plugin' is a valid plugin and load it
+	plugins, serverOpts, err := serverOptsFor(cmd)
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+
+	parallel, err := cmd.Flags().GetInt("parallel")
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+
+	all, err := cmd.Flags().GetBool("all")
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+
+	plug := args[0]
+	root, ok := plugins[plug]
+	if !ok {
+		// See if it's a script we can run as an external plugin instead
+		spec := plugin.ExternalPluginSpec{Script: plug}
+		root, err = spec.Load()
+		if err != nil {
+			pluginNames := make([]string, 0, len(plugins))
+			for name := range plugins {
+				pluginNames = append(pluginNames, name)
+			}
+			msg := "Unable to load %v as a configured plugin or script: %v\nValid plugins are: %v\n"
+			cmdutil.ErrPrintf(msg, plug, err, strings.Join(pluginNames, ", "))
+			return exitCode{1}
+		}
+	}
+
+	// Configure logging
+	log.SetFormatter(&log.TextFormatter{DisableTimestamp: true})
+	logFH, err := serverOpts.SetupLogging()
+	if err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+		return exitCode{1}
+	}
+	if logFH != nil {
+		defer logFH.Close()
+	}
+
+	registry := plugin.NewRegistry()
+	if err := registry.RegisterPlugin(root, serverOpts.PluginConfig[plug]); err != nil {
+		cmdutil.ErrPrintf("Error loading plugin: %v\n", err)
+		return exitCode{1}
+	}
+
+	rand.Seed(time.Now().UnixNano())
+	plugin.InitCache()
+	ctx := context.Background()
+
+	pw := progress.NewWriter()
+	pw.SetUpdateFrequency(50 * time.Millisecond)
+	pw.Style().Colors = progress.StyleColorsExample
+	go pw.Render()
+
+	// Use list to walk the hierarchy breadth-first. If a schema is present, visit each unique type
+	// at each level once, and stop when we've fully explored the schema. If there are errors, print
+	// all the detail we have about them. Use an unbuffered channel to register errors as they come in.
+	erred := 0
+	errs := make(chan error)
+	go func() {
+		for err := range errs {
+			erred++
+			cmdutil.ErrPrintf("%v\n", err)
+		}
+	}()
+
+	// We use a worker pool to limit work-in-progress, and a wait group to know when all queued work
+	// is complete. Put the plugin on the worker pool. Listing the registry should never error.
+	wp := newPool(parallel)
+	entries, err := plugin.CachedList(ctx, registry)
```

Comment:
A comment explaining why we're listing the registry instead of passing-in the root to processEntry would be nice. Guessing it's so the ID/other cached-list specific stuff is set?

Reply:
Yes. I'll add that.